[TECH] Rendre dynamique l'affichage des méthodes de connexions de type OIDC (PIX-5662)

igorissen opened the pull request:
## :unicorn: Problème

Actuellement, lors de l’ajout d’un nouveau partenaire de type OIDC, il faut également dupliquer du code au niveau de Pix Admin pour pouvoir supprimer la nouvelle méthode de connexion pour un utilisateur.

## :robot: Solution

Rendre l’affichage de la liste des méthodes de connexions de type OIDC dynamique sans devoir dupliquer beaucoup de code.

## :rainbow: Remarques

RAS

## :100: Pour tester

#### Seeds

- Mariamu Tatenda : 1 méthode de connexion (GAR)
- Paul Amplois : 2 méthodes de connexions (PIX;username + POLE_EMPLOI)
- Itai Chrizanne : 1 méthode de connexion (POLE_EMPLOI)
- David Cnav : 2 méthodes de connexions (PIX;username/email + CNAV)
- Ashura Onyinye : 1 méthode de connexion (CNAV)
- Charity Noble : Toutes les méthodes de connexions

#### Plusieurs méthodes de connexions

- Se connecter avec un compte ayant le rôle SUPERADMIN ou SUPPORT
- Ouvrir la page de détails d'un utilisateur ayant plusieurs méthodes de connexions incluant des méthodes de type OIDC
- Supprimer une méthode de connexion de type OIDC
- Constater qu'il n'y a pas de régression au niveau du processus de suppression (affichage d'une pop-up, valider la suppression, affichage d'une notification de succès)
- Déplacer une méthode de connexion de type OIDC vers un autre utilisateur
- Constater qu'il n'y a pas de régression au niveau du processus de déplacement (affichage d'une pop-up, choix de l'identifiant, valider le déplacement, affichage d'une notification de succès)

#### Administrateur n'ayant pas les droits de modification

- Se connecter avec un compte ayant le rôle CERTIF ou METIER
- Ouvrir la page de détails d'un utilisateur ayant plusieurs méthodes de connexions
- Constater qu'une grande partie des actions ne sont plus disponibles (supprimer, déplacer la méthode de connexion)

igorissen committed:
refactor(admin): move user getters

into authentication method component

igorissen committed:
feat(admin): display dynamically oidc providers

igorissen committed:
refactor(admin): update test

this case should not happen because the delete button appears only if the user has at least 2 authentication methods

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr4954.review.pix.fr
- App (.org): https://app-pr4954.review.pix.org
- Orga: https://orga-pr4954.review.pix.fr
- Certif: https://certif-pr4954.review.pix.fr
- Admin: https://admin-pr4954.review.pix.fr
- API: https://api-pr4954.review.pix.fr/api/

 Please check it out!

annemarie35 commented on admin/app/components/users/user-detail-personal-information/authentication-method.js:
```diff
@@ -6,13 +6,87 @@ import { inject as service } from '@ember/service';
   @tracked showAlreadyExistingEmailError = false;
+  @tracked selectedOidcAuthenticationMethod = null;
+
+  get hasPixAuthenticationMethod() {
```
Question : pourquoi les méthodes liées au modèle user ont été déplacées dans le composant ?

igorissen replied:
Ces méthodes ne sont utilisées que pour la gestion des méthodes d'authentification. Il n'était donc pas nécessaire de les avoir dans le modèle User (pollution 💨).
Et surtout c'est la 1ere étape nécessaire pour la suite du développement/mutualisation.

annemarie35 replied:
Ok, je sais pas, on pourrait dire aussi que c'est lié au modèle User, ce sont des règles métiers et comme ça on a tout au même endroit plutôt que dans un js d'un sous repertoire. Je ne sais pas ce qui est le mieux.
Quelle suite de mutualisation est prévue pour ma connaissance ?

igorissen replied:
Je comprends ton point de vue. Le modèle User peut être utilisé à plusieurs endroits ce qui veut dire que l'on aurait accès des règles métiers à des endroits où ce n'est pas nécessaire. Je vois ça plutôt comme ça en faite. Toute cette partie lié aux méthodes de connexion n'est utilisée qu'au niveau de ce composant.

Quand je parle de mutualisation en faite je veux parler du code qui permet de gérer la liste des méthodes de connexions de type OIDC. Toutes les autres méthodes de connexions resteront avec le même code séparé.

annemarie35 replied:
Merci pour le retour.
Je suis pas trop dispo pour explorer la chose en profondeur (surtout avec la magie faite par ember sur les modèles) mais il me semblait que côté admin on commencait déjà à avoir plusieurs "users" : les memberships (membre d'une orga), organizations learners, les admin members, etc. ce qui fait que user est un peu plus utilisé peut être.

annemarie35 commented on admin/tests/integration/components/users/user-detail-personal-information_test.js:
```diff
@@ -178,58 +176,17 @@ module('Integration | Component | users | user-detail-personal-information', fun
         assert.dom(screen.queryByRole('heading', { name: 'Merci de confirmer' })).doesNotExist();
         assert.dom(screen.queryByRole('button', { name: 'Annuler' })).doesNotExist();
         assert.dom(screen.queryByRole('button', { name: 'Confirmer' })).doesNotExist();
       });
```
Question : pourquoi avoir supprimer ce test  `should display an error message when the admin member try to remove the last authentication method`alors que le code `admin/app/components/users/user-detail-personal-information.js` qui génère cette erreur n'a pas été supprimé ?

igorissen replied:
J'ai oublié de supprimer aussi le code.

En faite il y a dans le composant authentication-method une vérification du nombre de méthodes de connexions permettant l'affichage ou non des boutons Supprimer. Vu que cette erreur n'est disponible que si l'on tente de supprimer la dernière méthode de connexion, j'ai trouvé impossible que ce cas arrive.

Quand on regarde le cas du test, il n'est pas logique. On ajoute 2 méthodes de connexions puis on en supprime 1 seule et on stub pour forcer l'erreur.

annemarie35 replied:
Je proposerais plutôt de remanier ce test qui vérifie toujours qu'**un admin ne peut pas supprimer une méthode de connexion quand l'utilisateur n'a plus qu'un seule méthode**. C'est un élément métier fort.
C'est impossible actuellement mais si on veut refactorer plus tard, on est plus serein car on saura si le test est toujours vert, qu'on a rien cassé.

igorissen replied:
True, je rajoute ce test manquant.

annemarie35 commented on admin/tests/unit/models/user_test.js:
```diff
@@ -27,118 +27,4 @@ module('Unit | Model | user', function (hooks) {
     });
   });
-
-  module('pix authentication method', function () {
```
Tristesse de voir disparaître des TU sur chaque méthode, tellement plus facile de savoir si on les a testé ou pas, snif ^^

igorissen replied:
😅

Ici plus vraiment besoin. Quelque soit la méthode de connexion de type OIDC, elles auront toutes le même comportement.

Maintenant si c'est nécessaire, on peut toujours faire une boucle sur un tableau de méthodes de connexion de type OIDC à tester.

annemarie35 replied:
Je voulais dire, on a déplacé ces méthodes dans `admin/app/components/users/user-detail-personal-information/authentication-method.js` mais y'a pas de TU de `authentication-method.js` 
Je sais pas quelle est la stratégie de tests mais j'ai remarqué qu'on fait pas souvent de TU pour ces methodes de controller

igorissen replied:
Ok. Vu qu'il y a des tests d'intégration, qui sont plus proche de l'utilisation d'un utilisateur, est-ce nécessaire d'avoir des TU ? Je comprends que les TU sont plus rapides à exécuter mais j'ai l'impression que les tests d'intégrations pour un composant c'est suffisant.

annemarie35 replied:
J'aime bien les TU dans ce cas car on peut tester plein de cas qui seraient plus longs à écrire en intégration (quand t'as par exemple 6 cas à tester)

annemarie35 commented on admin/app/components/users/user-detail-personal-information/authentication-method.js:
```diff
@@ -77,12 +145,22 @@ export default class AuthenticationMethod extends Component {
+    this.selectedOidcAuthenticationMethod = oidcAuthenticationMethod ? { ...oidcAuthenticationMethod } : null;
+    this.showReassignOidcAuthenticationMethodModal = !this.showReassignOidcAuthenticationMethodModal;
+
+    console.log(this.toggleAddAuthenticationMethodModal);
```
oubli ^^

igorissen replied:
🙈

annemarie35 commented on admin/app/components/users/user-detail-personal-information/authentication-method.js:
```diff
@@ -6,13 +6,87 @@ import { inject as service } from '@ember/service';
+      return {
+        code: oidcIdentityProvider.code,
+        name: oidcIdentityProvider.organizationName,
+        isActive: hasOidcAuthenticationMethod,
```
Je comprends le besoin de savoir quelles méthodes possède l'utilisateur par contre la notion de méthode "active" je trouve que ça induit en erreur : je me suis demandée c'était quoi une méthode inactive ^^ Dans la table des authentication-methods en BDD y'a pas disabledAt, quand on les supprime via l'admin, y'a plus les données en base.

les droits de l'admin sont dans le map via le canExecuteAction, ça me paraît un mélange de responsabilité

Pas de proposition de code à faire comme ça. 
Je dirais au moins un TU vu que ça mappe des données mais comme y'a this.args et this.accessControl je vois pas de manière rapide comment mocker ça.

igorissen replied:
Je comprends ton point.

Si je changes `isActive` en `hasOidcAuthenticationMethod` ou `hasAuthenticationMethod` ce serait mieux ?

Pour l'utilisation de `this.accessControl`, je vais le remettre au niveau du template et ajouter 2 nouvelles propriétés pour être plus précis, qui seront `canBeRemoved` et `canBeReassigned`. Est-ce que pour toi c'est ok comme cela ou tu as pu trouver une autre proposition ?

annemarie35 replied:
`hasAuthenticationMethod` c'est parfait !

yschoueri commented on admin/app/components/users/user-detail-personal-information/authentication-method.js:
```diff
@@ -6,13 +6,85 @@ import { inject as service } from '@ember/service';
+        canBeReassigned: hasAuthenticationMethod && this._hasMultipleAuthenticationMethods(),
+      };
+    });
+  }
```
🎉

yschoueri commented on admin/app/controllers/authenticated/users/get/information.js:
```diff
@@ -34,6 +34,11 @@ export default class UserInformationController extends Controller {
+    const oidcIdentityProvider = this.oidcIdentityProviders.list.findBy('code', identityProvider);
+    const reassignedAuthenticationMethodLabel = oidcIdentityProvider
+      ? oidcIdentityProvider.organizationName
+      : this.AUTHENTICATION_METHODS[identityProvider];
```
vu qu'il ne reste que le GAR dans `AUTHENTICATION_METHODS` est-ce qu'on en a encore besoin, ca ne pourrait pas etre mis direct ici ?

annemarie35 replied:
Ou alors, autre option, on prépare le refacto gar en saml en laissant tel quel mais en renommant `AUTHENTICATION_METHODS` en `SAML_AUTHENTICATION_METHODS`

yschoueri commented on admin/app/models/user.js:
```diff
@@ -36,72 +36,4 @@ export default class User extends Model {
-
-  get hasGarAuthenticationMethod() {
-    return this.authenticationMethods.any((authenticationMethod) => authenticationMethod.identityProvider === 'GAR');
-  }
-
-  get hasOnlyOneAuthenticationMethod() {
-    return (
-      [
-        this.hasEmailAuthenticationMethod,
-        this.hasUsernameAuthenticationMethod,
-        this.hasGarAuthenticationMethod,
-        this.hasPoleEmploiAuthenticationMethod,
-        this.hasCnavAuthenticationMethod,
-      ].filter((hasAuthenticationMethod) => hasAuthenticationMethod).length === 1
-    );
-  }
-
-  get isAllowedToRemoveEmailAuthenticationMethod() {
-    return this.hasEmailAuthenticationMethod && !this.hasOnlyOneAuthenticationMethod;
-  }
-
-  get isAllowedToRemoveUsernameAuthenticationMethod() {
-    return this.hasUsernameAuthenticationMethod && !this.hasOnlyOneAuthenticationMethod;
-  }
-
-  get isAllowedToRemovePoleEmploiAuthenticationMethod() {
-    return this.hasPoleEmploiAuthenticationMethod && !this.hasOnlyOneAuthenticationMethod;
-  }
-
-  get isAllowedToRemoveGarAuthenticationMethod() {
-    return this.hasGarAuthenticationMethod && !this.hasOnlyOneAuthenticationMethod;
-  }
-
-  get isAllowedToRemoveCnavAuthenticationMethod() {
-    return this.hasCnavAuthenticationMethod && !this.hasOnlyOneAuthenticationMethod;
-  }
-
-  get isAllowedToAddEmailAuthenticationMethod() {
-    return !this.hasPixAuthenticationMethod;
-  }
```
❤️ ❤️ ❤️

yschoueri commented on admin/tests/acceptance/authenticated/users/authentication-method_test.js:
```diff
@@ -91,33 +91,34 @@ module('Acceptance | authenticated/users | authentication-method', function (hoo
 
-  module('when reassign Pole Emploi authentication method', function () {
-    test('should remove Pole Emploi authentication method information', async function (assert) {
+  module('when reassign an oidc authentication method', function () {
```
```suggestion
  module('when reassigning an oidc authentication method', function () {
```

yschoueri commented on admin/tests/acceptance/authenticated/users/get_test.js:
```diff
@@ -113,10 +113,10 @@ module('Acceptance | authenticated/users/get', function (hooks) {
 
       const screen = await visit(`/users/${userToAnonymise.id}`);
-      await clickByName('Anonymiser cet utilisateur');
+      await click(screen.getByRole('button', { name: 'Anonymiser cet utilisateur' }));
```
❤️

igorissen committed:
refactor(api): move code

igorissen committed:
feat(api): update seeds

- add user with pole emploi identity provider
- add user with cnav identity provider
- add user with all providers

yschoueri commented:
Pour le user 100011 le bouton supprimer ne devrait pas apparaitre (voila les details sur dev) - il n'a qu'une connexion PE et rien d'autre 

![Screenshot 2022-09-22 at 17 28 18](https://user-images.githubusercontent.com/2871713/191788868-c3763dbc-82f9-4d43-8700-a08f1e4e8714.png)

annemarie35 commented on api/db/database-builder/factory/build-user.js:
```diff
@@ -47,6 +47,16 @@ function _buildPixAuthenticationMethod({
   });
 }
 
+function _generateAnEmailIfNecessary(email, id, lastName, firstName) {
```
Bizarre ^^

annemarie35 commented:
Trop cool pour les seeds, merci !
Paul Amplois : 2 méthodes de connexions (PIX + POLE_EMPLOI)
David Cnav : 2 méthodes de connexions (PIX + CNAV)
Charity Noble : Toutes les méthodes de connexions

Anne-Gaelle-S commented on admin/tests/unit/models/user_test.js:
```diff
@@ -27,118 +27,4 @@ module('Unit | Model | user', function (hooks) {
   });
-
-  module('pix authentication method', function () {
-    module('#hasPixAuthenticationMethod', function () {
```
**question**
Ces tests ont été supprimé mais pas déplacés ? Je me demande du coup s'il n'y a pas un trou dans la couverture de test... (je sais que ça va bouger avec les commits suivants alors je regarde la suite)

Anne-Gaelle-S replied:
Bon en effet je crois qu'on a quand même la couverture de test avec l'acceptance et l'integration...

Anne-Gaelle-S replied:
**chipotage**
Après je me demande si on aurait pas pu faire plus unitaire first... je sais que les tests d'acceptances étaient déjà en place. Mais du coup est-ce qu'on aurait pas pu faire comme pour l'API : uniquement le happy path en acceptance , le reste en unitaire et si c'est pas possible en integration 🤔

Anne-Gaelle-S reviewed (approved):
Il reste encore des occurences à "Pôle Emploi" et "CNAV" côté Pix Admin, je suggère qu'on créé un nouveau ticket pour traiter ça (plutôt que tout mettre dans cette PR).
